File: chapter2/app/app/src/main/java/net/numa08/kotlinbook/chapter2/viewmodels/ApplicationInformationViewModel.java
package net.numa08.kotlinbook.chapter2.viewmodels;

import android.content.Context;
import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.widget.ArrayAdapter;
import android.widget.BaseAdapter;

import net.numa08.kotlinbook.chapter2.BR;
import net.numa08.kotlinbook.chapter2.models.ApplicationInformation;
import net.numa08.kotlinbook.chapter2.repositories.ApplicationInformationRepository;

import java.util.ArrayList;

public final class ApplicationInformationViewModel extends BaseObservable {

    public interface Injector {
        Context getContext();
    }
    private final ApplicationInformationRepository applicationInformationRepository;

    @Nullable
    private ApplicationInformation applicationInformation;
    final private ArrayAdapter<String> adapter;
    private boolean isVisible;

    public ApplicationInformationViewModel(Injector injector, ApplicationInformationRepository applicationInformationRepository) {
        this.adapter = new ArrayAdapter<>(injector.getContext(), android.R.layout.simple_list_item_1, android.R.id.text1);
        this.applicationInformationRepository = applicationInformationRepository;
    }

    @Nullable @Bindable
    public ApplicationInformation getApplicationInformation() {
        return applicationInformation;
    }

    @Nullable @Bindable
    public Drawable getIcon() {
        if (applicationInformation == null) {
            return null;
        }
        return applicationInformation.getIcon();
    }

    private void setApplicationInformation(@NonNull final ApplicationInformation applicationInformation) {
        this.applicationInformation = applicationInformation;
        notifyPropertyChanged(BR.icon);
        notifyPropertyChanged(BR.applicationInformation);
        final ArrayList<String> list = new ArrayList<String>() {{
            if (applicationInformation.getApplicationInfo().className != null) {
                add(applicationInformation.getApplicationInfo().className);
            }
            if (applicationInformation.getApplicationInfo().backupAgentName != null) {
                add(applicationInformation.getApplicationInfo().backupAgentName);
            }
            if (applicationInformation.getApplicationInfo().dataDir != null) {
                add(applicationInformation.getApplicationInfo().dataDir);
            }
            if (applicationInformation.getApplicationInfo().manageSpaceActivityName != null) {
                add(applicationInformation.getApplicationInfo().manageSpaceActivityName);
            }
            if (applicationInformation.getApplicationInfo().nativeLibraryDir != null) {
                add(applicationInformation.getApplicationInfo().nativeLibraryDir);
            }
            if (applicationInformation.getApplicationInfo().permission != null) {
                add(applicationInformation.getApplicationInfo().permission);
            }
            if (applicationInformation.getApplicationInfo().publicSourceDir != null) {
                add(applicationInformation.getApplicationInfo().publicSourceDir);
            }
            if (applicationInformation.getApplicationInfo().sourceDir != null) {
                add(applicationInformation.getApplicationInfo().sourceDir);
            }
            if (applicationInformation.getApplicationInfo().taskAffinity != null) {
                add(applicationInformation.getApplicationInfo().taskAffinity);
            }
        }};
        adapter.addAll(list);
    }

    public BaseAdapter getAdapter() {
        return adapter;
    }

    public void onCreate() {
        isVisible = true;
    }

    public void onDestroy() {
        isVisible = false;
    }

    public void fetchApplication(String packageName) {
        applicationInformationRepository.findApplicationByPackageName(packageName, new ApplicationInformationRepository.FindApplicationCallback() {
            @Override
            public void onFindApplication(@Nullable ApplicationInformation information) {
                if (isVisible && information != null) {
                    setApplicationInformation(information);
                }
            }
        });
    }
}

File: chapter2/app/app/src/main/java/net/numa08/kotlinbook/chapter2/viewmodels/ApplicationInformationViewModel.kt
package net.numa08.kotlinbook.chapter2.viewmodels

import android.content.Context
import android.databinding.BaseObservable
import android.databinding.Bindable
import android.graphics.drawable.Drawable
import android.widget.ArrayAdapter
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch
import net.numa08.kotlinbook.chapter2.BR
import net.numa08.kotlinbook.chapter2.models.ApplicationInformation
import net.numa08.kotlinbook.chapter2.repositories.ApplicationInformationRepository

class ApplicationInformationViewModel(injector: Injector, private val applicationInformationRepository: ApplicationInformationRepository) : BaseObservable() {

    @get:Bindable
    var applicationInformation: ApplicationInformation? = null
        private set(applicationInformation) {
            field = applicationInformation
            notifyPropertyChanged(BR.icon)
            notifyPropertyChanged(BR.applicationInformation)
            val list = applicationInformation?.applicationInfo?.let {
                listOf(
                        it.className,
                        it.backupAgentName,
                        it.dataDir,
                        it.manageSpaceActivityName,
                        it.nativeLibraryDir,
                        it.permission,
                        it.publicSourceDir,
                        it.sourceDir,
                        it.taskAffinity
                ).filterNotNull()
            }
            if (list != null) {
                adapter.addAll(list)
            }
        }
    val adapter: ArrayAdapter<String> = ArrayAdapter(injector.context, android.R.layout.simple_list_item_1, android.R.id.text1)

    var isVisible: Boolean = false
        private set

    var job: Job? = null
        private set

    val icon: Drawable?
        @Bindable get() = applicationInformation?.icon

    interface Injector {
        val context: Context
    }

    fun onCreate() {
        isVisible = true
    }

    fun onDestroy() {
        isVisible = false
        job?.cancel()
    }

    fun fetchApplication(packageName: String) {
        job = launch(UI) {
            val info = applicationInformationRepository.findApplicationByPackageNameAsync(packageName).await()
            if (isVisible && info != null) {
                applicationInformation = info
            }
        }
    }
}
